design/design-doc-aggregate-resources.md

```diff
@@ -0,0 +1,965 @@
+
+This document proposes that `AggregateResource` connection secrets contain the
+intersection of the connection secrets of the managed resources they aggregate. 
+For example if an `AggregateResource` matched two managed resources with the
+following secrets:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: secret-a
+stringData:
+  username: cooluser
+  password: verysecure
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: secret-b
+stringData:
+  url: https://example.org
+```
+
+The `AggregateResource` connection secret would be:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: aggregate-secret
+stringData:
+  username: cooluser
+  password: verysecure
+  url: https://example.org
+```
+
+An undefined connection secret key will win in the case of a conflict. Conflicts
+are expected to be rare in practice (most "supporting" managed resources do not
+create a connection secret), but may be avoided by respecting a new
+`aggregation.crossplane.io/connection-secret-prefix` annotation. This annotation
+would be propagated from resource classes to the managed resources they
+dynamically provision, specifying a string to be prepended to their connection
+secret keys during aggregation.
+
+## User Experience
+
+Consider an infrastructure operator who wishes to enable their application
+operator peers to submit `MySQLInstance` resource claims and have them be
+satisfied by dynamically provisioned Azure `MySQLServer` managed resources. Each
+`MySQLServer` should allow traffic from a statically provisioned subnetwork
+using a `MySQLServerVirtualNetworkRule`. Each dynamically provisioned
+`MySQLServer` and its associated virtual network rule should be created in its
+own resource group.
+
+First the infrastructure operator statically provisions the `Subnet` from which
+all dynamically provisioned `MySQLServer` managed resources should allow
+traffic:
+
+```yaml
+apiVersion: network.azure.crossplane.io/v1alpha3
+kind: Subnet
+metadata:
+  name: private
+spec:
+  resourceGroupName:
+    name: some-existing-group
+  virtualNetworkName:
+    name: some-existing-network
+  properties:
+    addressPrefix: 10.2.0.0/24
+    serviceEndpoints:
+      - service: Microsoft.Sql
+  providerRef:
+    name: example
+```
+
+The infrastructure operator then authors resource classes for each of the
+managed resources they wish to allow application operators to dynamically
+provision, and an aggregate resource class to aggregate them:
+
+```yaml
+apiVersion: aggregation.crossplane.io/v1alpha1
+kind: AggregateResourceClass
+metadata:
+  name: private-mysql-server
+  annotations:
+    # This will be the default resource class for any resource claim capable of
+    # binding directly to a resource of one of the classes it aggregates - in
+    # this case MySQLInstance.
+    resourceclass.crossplane.io/is-default-class: "true"
```

Comment:
> So, the cost of the random selection is highly different with AggregateResourceClass compared to other individual resource classes.

I just pushed a commit that hopefully clarifies this a little. `AggregateResourceClass` can _match_ arbitrary classes, but it will only be processed by controllers that are _concerned with_ those classes, similar to how the `resource.Claim*Reconciler` logic works today. The stack author will choose which resources may aggregate to which claims - we don't allow an arbitrary mapping.

Reply:
So, let's say I have 1 `NetworkClass`, 1 `SubnetworkClass`, 1 `GlobalAddressClass` and 1 `GKEClusterClass` and 1 `CloudSQLClass`. When I create a `KubernetesCluster` claim that refers to this `AggregateClass` that includes all these classes; I will get 1 `Network`, `Subnetwork`, `GlobalAddress` and `GKECluster` but no `CloudSQLInstance`, right?

Reply:
Assuming there are controllers that know how to produce a Network, Subnetwork GlobalAddress, and GKECluster from a KubernetesCluster, yes.